WSDL url reference from bitbucket doesnt work in cli but it works in api manager UI

### APIM-CLI version

1.10.1

### API-Management version

7.7.20220228

### Question

Hi Chris ,

I am working on wsdl service on boarding , my use case is to onboard the wsdl service using url , since apimg doesnt support wsdl file . i am storing the wsdl file in bitbucket so that i can use the bitbucket refernce url in apimanager to onboard it works .but when i use the same url through apimg cli but it throws error saying invalid api-sec

my query is the usernameand password for bit bucket i am passing in below format

username/password@https://bitbucketurl .
[username/password@]https://bitbucketurl .

The observation is this bitbucket url has query params . the same url with query params i tried in api manager with bit bucket credentials it works but doesnt work on apimg cli . 

could you let me know how this apimcli uses the url i provided , what is the right way to pass the credentials ?

i have opened axway support ticket Case 01358292 as well . but the support team referred me to report it on git hub

Regards
Tabrez

Hi Chris , 

Sorry for delay to get back . below is the debug mode logs andthough i am passing bit bucket credentials , the url is just routing to bitbucket login page , where in when i pass the same url in apimanager . it extracts the api-specification. 

openjdk version "1.8.0-292"
OpenJDK Runtime Environment (build 1.8.0-292-b10)
OpenJDK 64-Bit Server VM (build 25.71-b10, mixed mode)

20220418-094238
Taking Backup of Existing APIMG Swagger files
int-manager/
int-manager/AEM-ReadExtension/
int-manager/AEM-ReadExtension/swagger/
int-manager/AEM-ReadExtension/swagger/ReadExtension.url
int-manager/AEM-ReadExtension/config/
int-manager/AEM-ReadExtension/config/MAS-APIKEY_Inbound-NoAuth_Outbound.json
copied backup to /data/mas/backup/int-manager-bkp-20220418-094238.tar.gz
Apimg promotion start
<--------------- In AEM-ReadExtension directory -------------------------->

----------------------------------------------------------------------------------------
API-Manager CLI: 1.10.1

To report issues or get help, please visit:
https://github.com/Axway-API-Management-Plus/apim-cli
----------------------------------------------------------------------------------------
Module: API - I M P O R T (1.10.1)
----------------------------------------------------------------------------------------
2756   DEBUG nvironmentProperties| Trying to load environment properties from file: /data/Software/apim-cli-1.10.1/conf/env.properties ... not found.
2778   DEBUG nvironmentProperties| Loaded environment properties from file: /data/Software/apim-cli-1.10.1/conf/env.api-dev.properties
2943   DEBUG    APIManagerAdapter| Logging in with User: 'apiadmin'
7416   DEBUG ManagerConfigAdapter| Load API-Manager configuration.
7667   INFO     APIManagerAdapter| Successfully connected to API-Manager (7.7.20220228) on: https://XXXXXXXXXX:8075
7771   DEBUG rOrganizationAdapter| Load organizations from API-Manager using filter: OrgFilter [name=MAS-Internal, id=null]
8864   DEBUG nagerPoliciesAdapter| Load policies with type: REQUEST from API-Manager
10727  DEBUG nagerPoliciesAdapter| Load policies with type: RESPONSE from API-Manager
10882  INFO                 Utils| Loading API-Definition from: https://git.mas.gov.sg/projects/APIMG/repos/apigw-utils/raw/wsdls/BarcodedForms.wsdl?at=refs%2Fheads%2Fmaster (c_lesterchan)
10884  DEBUG                Utils| NO javax.net.ssl.keyStore property.
10886  DEBUG                Utils| https protocols: TLSv1.2
11012  DEBUG SpecificationFactory| Handle API-Specification: '<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><title>Log in - Bitbucket</title><script>
window.WRM=window.WRM||{};window.WRM._unparsed......', apiDefinitionFile: '/data/mas/apimg/int-manager/AEM-ReadExtension/swagger/ReadExtension.url'
11023  DEBUG SpecificationFactory| Can't handle API specification with class: com.axway.apim.api.apiSpecification.Swagger2xSpecification
11024  DEBUG SpecificationFactory| Can't handle API specification with class: com.axway.apim.api.apiSpecification.Swagger1xSpecification
11024  DEBUG SpecificationFactory| Can't handle API specification with class: com.axway.apim.api.apiSpecification.OAS3xSpecification
11024  DEBUG    WSDLSpecification| No WSDL specification. Specification doesn't contain wsdl in the first 500 characters.
11024  DEBUG SpecificationFactory| Can't handle API specification with class: com.axway.apim.api.apiSpecification.WSDLSpecification
11025  DEBUG    WADLSpecification| No WADL specification. Specification doesn't contain WADL namespace: wadl.dev.java.net in the first 500 characters.
11025  DEBUG SpecificationFactory| Can't handle API specification with class: com.axway.apim.api.apiSpecification.WADLSpecification
11111  ERROR ODataV2Specification| Error parsing OData specification.
org.apache.olingo.odata2.api.ep.EntityProviderException: An exception of type 'WstxUnexpectedCharException' occurred.
        at org.apache.olingo.odata2.core.ep.consumer.XmlMetadataConsumer.readMetadata(XmlMetadataConsumer.java:123) ~[olingo-odata2-core-2.0.11.jar:2.0.11]
        at org.apache.olingo.odata2.core.edm.provider.EdmxProvider.parse(EdmxProvider.java:51) ~[olingo-odata2-core-2.0.11.jar:2.0.11]
        at org.apache.olingo.odata2.core.ep.ProviderFacadeImpl.readMetadata(ProviderFacadeImpl.java:231) ~[olingo-odata2-core-2.0.11.jar:2.0.11]
        at org.apache.olingo.odata2.api.ep.EntityProvider.readMetadata(EntityProvider.java:876) ~[olingo-odata2-api-2.0.11.jar:2.0.11]
        at com.axway.apim.api.apiSpecification.ODataV2Specification.parse(ODataV2Specification.java:91) [apimcli-apim-adapter-1.10.1.jar:1.10.1]
        at com.axway.apim.api.apiSpecification.APISpecificationFactory.getAPISpecification(APISpecificationFactory.java:77) [apimcli-apim-adapter-1.10.1.jar:1.10.1]
        at com.axway.apim.api.apiSpecification.APISpecificationFactory.getAPISpecification(APISpecificationFactory.java:57) [apimcli-apim-adapter-1.10.1.jar:1.10.1]
        at com.axway.apim.apiimport.APIImportConfigAdapter.addAPISpecification(APIImportConfigAdapter.java:231) [apimcli-apis-1.10.1.jar:1.10.1]
        at com.axway.apim.apiimport.APIImportConfigAdapter.getDesiredAPI(APIImportConfigAdapter.java:179) [apimcli-apis-1.10.1.jar:1.10.1]
        at com.axway.apim.APIImportApp.importAPI(APIImportApp.java:76) [apimcli-apis-1.10.1.jar:1.10.1]
        at com.axway.apim.APIImportApp.importAPI(APIImportApp.java:58) [apimcli-apis-1.10.1.jar:1.10.1]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0-292]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0-292]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0-292]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0-292]
        at com.axway.apim.cli.APIManagerCLI.run(APIManagerCLI.java:141) [apimcli-tool-1.10.1.jar:1.10.1]
        at com.axway.apim.cli.APIManagerCLI.main(APIManagerCLI.java:77) [apimcli-tool-1.10.1.jar:1.10.1]
Caused by: com.ctc.wstx.exc.WstxUnexpectedCharException: Unexpected character '>' (code 62) expected '='
 at [row,col {unknown-source}]: [15,269]
        at com.ctc.wstx.sr.StreamScanner.throwUnexpectedChar(StreamScanner.java:666) ~[woodstox-core-6.2.6.jar:6.2.6]
        at com.ctc.wstx.sr.BasicStreamReader.handleNsAttrs(BasicStreamReader.java:3117) ~[woodstox-core-6.2.6.jar:6.2.6]
        at com.ctc.wstx.sr.BasicStreamReader.handleStartElem(BasicStreamReader.java:3048) ~[woodstox-core-6.2.6.jar:6.2.6]
        at com.ctc.wstx.sr.BasicStreamReader.nextFromTree(BasicStreamReader.java:2926) ~[woodstox-core-6.2.6.jar:6.2.6]
        at com.ctc.wstx.sr.BasicStreamReader.next(BasicStreamReader.java:1122) ~[woodstox-core-6.2.6.jar:6.2.6]
        at org.apache.olingo.odata2.core.ep.consumer.XmlMetadataConsumer.readMetadata(XmlMetadataConsumer.java:97) ~[olingo-odata2-core-2.0.11.jar:2.0.11]
        ... 16 more
11131  DEBUG SpecificationFactory| Can't handle API specification with class: com.axway.apim.api.apiSpecification.ODataV2Specification
11131  DEBUG SpecificationFactory| Can't handle API specification with class: com.axway.apim.api.apiSpecification.ODataV3Specification
11131  DEBUG SpecificationFactory| Can't handle API specification with class: com.axway.apim.api.apiSpecification.ODataV4Specification
11131  ERROR SpecificationFactory| API: 'AEM-Read Extension Services' has a unkown/invalid API-Specification: '<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><title>Log in - Bitbucket</title><script>
window.WRM=window.WRM||{};window.WRM._unparsed...'
11134  ERROR         AppException| Can't handle API specification. No suiteable API-Specification implementation available.
com.axway.apim.lib.errorHandling.AppException: Can't handle API specification. No suiteable API-Specification implementation available.
        at com.axway.apim.api.apiSpecification.APISpecificationFactory.getAPISpecification(APISpecificationFactory.java:103) ~[apimcli-apim-adapter-1.10.1.jar:1.10.1]
        at com.axway.apim.api.apiSpecification.APISpecificationFactory.getAPISpecification(APISpecificationFactory.java:57) ~[apimcli-apim-adapter-1.10.1.jar:1.10.1]
        at com.axway.apim.apiimport.APIImportConfigAdapter.addAPISpecification(APIImportConfigAdapter.java:231) ~[apimcli-apis-1.10.1.jar:1.10.1]
        at com.axway.apim.apiimport.APIImportConfigAdapter.getDesiredAPI(APIImportConfigAdapter.java:179) ~[apimcli-apis-1.10.1.jar:1.10.1]
        at com.axway.apim.APIImportApp.importAPI(APIImportApp.java:76) [apimcli-apis-1.10.1.jar:1.10.1]
        at com.axway.apim.APIImportApp.importAPI(APIImportApp.java:58) [apimcli-apis-1.10.1.jar:1.10.1]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0-292]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0-292]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0-292]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0-292]
        at com.axway.apim.cli.APIManagerCLI.run(APIManagerCLI.java:141) [apimcli-tool-1.10.1.jar:1.10.1]
        at com.axway.apim.cli.APIManagerCLI.main(APIManagerCLI.java:77) [apimcli-tool-1.10.1.jar:1.10.1]
11134  DEBUG    APIManagerAdapter| Closing cache ...

--- a/modules/apim-adapter/src/main/java/com/axway/apim/adapter/apis/APIManagerAPIAdapter.java
+++ b/modules/apim-adapter/src/main/java/com/axway/apim/adapter/apis/APIManagerAPIAdapter.java
@@ -60,6 +60,7 @@ import com.axway.apim.api.model.apps.ClientApplication;
 import com.axway.apim.lib.CoreParameters;
 import com.axway.apim.lib.errorHandling.AppException;
 import com.axway.apim.lib.errorHandling.ErrorCode;
+import com.axway.apim.lib.utils.URLParser;
 import com.axway.apim.lib.utils.Utils;
 import com.axway.apim.lib.utils.rest.DELRequest;
 import com.axway.apim.lib.utils.rest.GETRequest;
@@ -873,9 +874,10 @@ public class APIManagerAPIAdapter {
 		} else {
 			completeWsdlUrl = api.getApiDefinition().getApiSpecificationFile();
 		}
-		wsdlUrl = extractURI(completeWsdlUrl);
-		username=extractUsername(completeWsdlUrl);
-		pass=extractPassword(completeWsdlUrl);
+		URLParser parser = new URLParser(completeWsdlUrl);
+		wsdlUrl = parser.getUri();
+		username = parser.getUsername();
+		pass = parser.getPassword();
 
 		try {
 			URIBuilder uriBuilder = new URIBuilder(cmd.getAPIManagerURL()).setPath(cmd.getApiBasepath()+"/apirepo/importFromUrl/")
@@ -1140,34 +1142,6 @@ public class APIManagerAPIAdapter {
 		}
 		filterFields += "]";
 		return filterFields;
-	}
-	
-	private String extractUsername(String url) {
-		String[] temp = url.split("@");
-		if(temp.length==2) {
-			return temp[0].substring(0, temp[0].indexOf("/"));
-		}
-		return null;
-	}
-	
-	private String extractPassword(String url) {
-		String[] temp = url.split("@");
-		if(temp.length==2) {
-			return temp[0].substring(temp[0].indexOf("/")+1);
-		}
-		return null;
-	}
-	
-	private String extractURI(String url) throws AppException
-	{
-		String[] temp = url.split("@");
-		if(temp.length==1) {
-			return temp[0];
-		} else if(temp.length==2) {
-			return temp[1];
-		} else {
-			throw new AppException("WSDL-URL has an invalid format. ", ErrorCode.CANT_READ_WSDL_FILE);
-		}
 	}
 	
 	public APIManagerAPIAdapter setAPIManagerResponse(APIFilter filter, String apiManagerResponse) {

--- a/modules/apim-adapter/src/main/java/com/axway/apim/api/apiSpecification/APISpecificationFactory.java
+++ b/modules/apim-adapter/src/main/java/com/axway/apim/api/apiSpecification/APISpecificationFactory.java
@@ -12,13 +12,9 @@ import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 
 import org.apache.commons.io.IOUtils;
-import org.apache.http.HttpEntity;
-import org.apache.http.HttpResponse;
-import org.apache.http.client.ClientProtocolException;
-import org.apache.http.client.ResponseHandler;
 import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.CloseableHttpResponse;
 import org.apache.http.client.methods.HttpGet;
-import org.apache.http.impl.client.CloseableHttpClient;
 import org.apache.http.util.EntityUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -26,6 +22,7 @@ import org.slf4j.LoggerFactory;
 import com.axway.apim.api.model.DesiredAPISpecification;
 import com.axway.apim.lib.errorHandling.AppException;
 import com.axway.apim.lib.errorHandling.ErrorCode;
+import com.axway.apim.lib.utils.HTTPClient;
 import com.axway.apim.lib.utils.URLParser;
 import com.axway.apim.lib.utils.Utils;
 
@@ -159,7 +156,10 @@ public class APISpecificationFactory {
 		String uri = url.getUri();
 		String username = url.getUsername();
 		String password = url.getPassword();
-		CloseableHttpClient httpclient = Utils.createHttpClient(uri, username, password);
+		CloseableHttpResponse httpResponse = null;
+		
+		HTTPClient httpClient = new HTTPClient(uri, username, password);
+
 		try {
 			RequestConfig config = RequestConfig.custom()
 					.setRelativeRedirectsAllowed(true)
@@ -167,29 +167,22 @@ public class APISpecificationFactory {
 					.build();
 			HttpGet httpGet = new HttpGet(uri);
 			httpGet.setConfig(config);
-			
-            ResponseHandler<String> responseHandler = new ResponseHandler<String>() {
 
-                @Override
-                public String handleResponse(
-                        final HttpResponse response) throws ClientProtocolException, IOException {
-                    int status = response.getStatusLine().getStatusCode();
-                    if (status >= 200 && status < 300) {
-                        HttpEntity entity = response.getEntity();
-                        return entity != null ? EntityUtils.toString(entity,StandardCharsets.UTF_8) : null;
-                    } else {
-                        throw new ClientProtocolException("Unexpected response status: " + status);
-                    }
-                }
-
-            };
-            String responseBody = httpclient.execute(httpGet, responseHandler);
-            return new ByteArrayInputStream(responseBody.getBytes(StandardCharsets.UTF_8));
+            httpResponse = httpClient.execute(httpGet);
+            int statusCode = httpResponse.getStatusLine().getStatusCode();
+            String response = EntityUtils.toString(httpResponse.getEntity(), StandardCharsets.UTF_8);
+            if (statusCode >= 200 && statusCode < 300) {
+            	return new ByteArrayInputStream(response.getBytes(StandardCharsets.UTF_8));
+            } else {
+            	throw new AppException("Cannot load API-Specification from URI. Received Status-Code: " + statusCode + ", Response: '" + response + "'", 
+            			ErrorCode.CANT_READ_API_DEFINITION_FILE);
+            }
 		} catch (Exception e) {
-			throw new AppException("Cannot load API-Definition from URI: "+uri, ErrorCode.CANT_READ_API_DEFINITION_FILE, e);
+			throw new AppException("Cannot load API-Specification from URI: "+uri, ErrorCode.CANT_READ_API_DEFINITION_FILE, e);
 		} finally {
 			try {
-				httpclient.close();
+				httpClient.close();
+				httpResponse.close();
 			} catch (Exception ignore) {}
 		}
 	}

--- a/modules/apim-adapter/src/main/java/com/axway/apim/lib/utils/HTTPClient.java
+++ b/modules/apim-adapter/src/main/java/com/axway/apim/lib/utils/HTTPClient.java
@@ -1,0 +1,93 @@
+package com.axway.apim.lib.utils;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.security.KeyManagementException;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+
+import org.apache.http.HttpHost;
+import org.apache.http.auth.AuthScope;
+import org.apache.http.auth.UsernamePasswordCredentials;
+import org.apache.http.client.AuthCache;
+import org.apache.http.client.CredentialsProvider;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.conn.ssl.NoopHostnameVerifier;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
+import org.apache.http.conn.ssl.TrustAllStrategy;
+import org.apache.http.impl.auth.BasicScheme;
+import org.apache.http.impl.client.BasicAuthCache;
+import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.ssl.SSLContextBuilder;
+
+import com.axway.apim.lib.errorHandling.AppException;
+import com.axway.apim.lib.errorHandling.ErrorCode;
+
+public class HTTPClient {
+	
+	private URI url;
+	private String password;
+	private String username;
+	
+	private CloseableHttpClient httpClient = null;
+	
+	private HttpClientContext clientContext;
+	
+	public HTTPClient(String url, String username, String password) throws AppException {
+		super();
+		try {
+			this.url = new URI(url);
+			this.password = password;
+			this.username = username;
+			getClient();
+		} catch (URISyntaxException e) {
+			throw new AppException("Error creating HTTP-Client.", ErrorCode.UNXPECTED_ERROR, e);
+		}
+	}
+
+	public void getClient() throws AppException {
+		try {
+			SSLContextBuilder builder = SSLContextBuilder.create();
+			builder.loadTrustMaterial(null, new TrustAllStrategy());
+			SSLConnectionSocketFactory sslsf = new SSLConnectionSocketFactory(builder.build(), new NoopHostnameVerifier());
+			
+			HttpClientBuilder httpClientBuilder = HttpClients.custom()
+					.setSSLSocketFactory(sslsf);
+			
+			if(this.username!=null) {
+				CredentialsProvider credsProvider = new BasicCredentialsProvider();
+				credsProvider.setCredentials(AuthScope.ANY, new UsernamePasswordCredentials(username, password));
+				AuthCache authCache = new BasicAuthCache();
+				BasicScheme basicAuth = new BasicScheme();
+				authCache.put( new HttpHost(url.getHost(), url.getPort(), url.getScheme()), basicAuth );
+				clientContext = HttpClientContext.create();
+				clientContext.setAuthCache(authCache);
+	
+				httpClientBuilder.setDefaultCredentialsProvider(credsProvider);
+			}
+		
+			this.httpClient = httpClientBuilder.build();
+		} catch (NoSuchAlgorithmException | KeyStoreException | KeyManagementException e) {
+			throw new AppException("Error creating HTTP-Client.", ErrorCode.UNXPECTED_ERROR, e);
+		}
+	}
+	
+	public CloseableHttpResponse execute(HttpUriRequest request) throws Exception {
+		CloseableHttpResponse response = httpClient.execute(request, clientContext);
+		return response;
+	}
+	
+	public void close() {
+		try {
+			this.httpClient.close();
+		} catch (IOException e) {
+		}
+	}
+
+}

--- a/modules/apim-adapter/src/main/java/com/axway/apim/lib/utils/Utils.java
+++ b/modules/apim-adapter/src/main/java/com/axway/apim/lib/utils/Utils.java
@@ -323,42 +323,6 @@ public class Utils {
 		return api.getName() + " " + api.getVersion() + " ("+api.getVersion()+")";
 	}
 	
-	public static CloseableHttpClient createHttpClient(String uri, String username, String password) throws AppException {
-		HttpClientBuilder httpClientBuilder = HttpClients.custom();
-		try {
-			addBasicAuthCredential(uri, username, password, httpClientBuilder);
-			addSSLContext(uri, httpClientBuilder);
-			return httpClientBuilder.build();
-		} catch (Exception e) {
-			throw new AppException("Error during create http client for retrieving ...", ErrorCode.CANT_CREATE_HTTP_CLIENT);
-		}
-	}
-
-	private static void addSSLContext(String uri, HttpClientBuilder httpClientBuilder) throws KeyManagementException,
-			NoSuchAlgorithmException, KeyStoreException, CertificateException, IOException, UnrecoverableKeyException {
-		if (isHttpsUri(uri)) {
-			SSLConnectionSocketFactory sslCtx = createSSLContext();
-			if (sslCtx!=null) {
-				httpClientBuilder.setSSLSocketFactory(sslCtx);
-			}
-		}
-	}
-
-	private static void addBasicAuthCredential(String uri, String username, String password,
-			HttpClientBuilder httpClientBuilder) {
-		//if(this.apiConfig instanceof DesiredTestOnlyAPI) return; // Don't do that when unit-testing
-		if(username!=null) {
-			LOG.info("Loading API-Definition from: " + uri + " ("+username+")");
-			CredentialsProvider credsProvider = new BasicCredentialsProvider();
-			credsProvider.setCredentials(
-		            new AuthScope(AuthScope.ANY),
-		            new UsernamePasswordCredentials(username, password));
-			httpClientBuilder.setDefaultCredentialsProvider(credsProvider);
-		} else {
-			LOG.info("Loading API-Definition from: " + uri);
-		}
-	}
-	
 	public static boolean isHttpUri(String pathToAPIDefinition) {
 		String httpUri = pathToAPIDefinition.substring(pathToAPIDefinition.indexOf("@")+1);
 		return( httpUri.startsWith("http://") || httpUri.startsWith("https://"));
@@ -366,37 +330,5 @@ public class Utils {
 	
 	public static boolean isHttpsUri(String uri) {
 		return( uri.startsWith("https://") );
-	}
-	
-	private static SSLConnectionSocketFactory createSSLContext() throws KeyManagementException, NoSuchAlgorithmException, KeyStoreException, CertificateException, IOException, UnrecoverableKeyException {
-		SSLContextBuilder builder = new SSLContextBuilder();
-		builder.loadTrustMaterial(null, new TrustAllStrategy());
-		
-		String keyStorePath=System.getProperty("javax.net.ssl.keyStore","");
-		if (StringUtils.isNotEmpty(keyStorePath)) {
-			String keyStorePassword=System.getProperty("javax.net.ssl.keyStorePassword","");
-			if (StringUtils.isNotEmpty(keyStorePassword)) {
-				String keystoreType=System.getProperty("javax.net.ssl.keyStoreType",KeyStore.getDefaultType());
-				LOG.debug("Reading keystore from {}",keyStorePath);
-				KeyStore ks = KeyStore.getInstance(keystoreType);
-				ks.load(new FileInputStream(new File(keyStorePath)), keyStorePassword.toCharArray());				
-				builder.loadKeyMaterial(ks,keyStorePassword.toCharArray());
-			}
-		} else {
-			LOG.debug("NO javax.net.ssl.keyStore property.");
-		}
-		String [] tlsProts = getAcceptedTLSProtocols();
-		SSLConnectionSocketFactory sslsf = new SSLConnectionSocketFactory(
-				builder.build(),
-                tlsProts,
-                null,
-                new NoopHostnameVerifier());
-		return sslsf;
-	}
-
-	private static String[] getAcceptedTLSProtocols() {
-		String protocols = System.getProperty("https.protocols","TLSv1.2"); //default TLSv1.2
-		LOG.debug("https protocols: {}",protocols);
-		return protocols.split(",");
 	}
 }

--- a/modules/apim-adapter/src/test/java/com/axway/lib/utils/URLParserTest.java
+++ b/modules/apim-adapter/src/test/java/com/axway/lib/utils/URLParserTest.java
@@ -44,4 +44,12 @@ public class URLParserTest {
 		String urlToAPIDefinition = "dadasdsa@https://petstore.swagger.io/v2/swagger.json";
 		assertThrows(AppException.class, () -> {new URLParser(urlToAPIDefinition);});
 	}
+	
+	@Test
+	public void testPasswordWithAt() throws AppException {
+		String urlToAPIDefinition = "user@axway.com/abc@12345@https://petstore.swagger.io/v2/swagger.json";
+		URLParser parser = new URLParser(urlToAPIDefinition);
+		Assert.assertEquals(parser.getUsername(), "user@axway.com");
+		Assert.assertEquals(parser.getPassword(), "abc@12345");
+	}
 }

